source/Octopus.Tentacle/Kubernetes/Synchronisation/Internal/SemaphoreSlimReleaser.cs

```diff
@@ -0,0 +1,24 @@
+using System;
+using System.Threading;
+
+namespace Octopus.Tentacle.Kubernetes.Synchronisation.Internal
+{
+    class SemaphoreSlimReleaser<T> : ISemaphoreSlimReleaser<T> where T : SemaphoreSlim
```

Comment:
Rather than having a Factory and specific interface for this, I'd probably make it a private class in `ReferenceCountingKeyedLock`. Maybe same with `SemaphoreSlimWithReferenceCount`.

Reply:
I'll see what I can do to reduce the number of classes/interfaces. I only added the Factory in order to inject some listeners for testing the dispose callback.

Reply:
Ah yep fair enough. You might be able to introduce a seam instead 🤔